java/org/openobservatory/measurement_kit/nettests/OoniTestBase.java

```diff
@@ -50,25 +66,71 @@ public OoniTestBase use_logcat() {
     }
 
     public OoniTestBase on_log(LogCallback delegate) {
-        wrapper.on_log(delegate);
+        this.logDelegate = delegate;
         return this;
     }
 
     public OoniTestBase on_entry(EntryCallback delegate) {
-        wrapper.on_entry(delegate);
+        this.entryDelegate = delegate;
         return this;
     }
 
-    public void run(TestCompleteCallback callback) {
-        wrapper.run(callback);
-    }
+
 
     public OoniTestBase set_options(String key, String value) {
         wrapper.set_options(key, value);
         return this;
     }
 
-    public void run() {
-        wrapper.run();
+    public Integer run(TestCompleteCallback callback) {
+        this.run_impl(callback);
+        return testId;
     }
+
+    public Integer run() {
+        // XXX: it's actually not safe to use a blocking callback on Android
+        // If the UI Thread is not responding for more than 3 seconds the system kills the app
```

Comment:
What stated in the comment above is true. However, the use case of the synchronous run callback is that the library user would create a thread or an async task and then call run(). Thus, I believe the comment should be phrased differently to explicitly include this possibility.